modules/runner/src/main/java/org/apache/ignite/internal/storage/LocalConfigurationStorage.java

```diff
@@ -49,47 +69,121 @@ public LocalConfigurationStorage(VaultManager vaultMgr) {
         this.vaultMgr = vaultMgr;
     }
 
-    /** Map to store values. */
-    private Map<String, Serializable> map = new ConcurrentHashMap<>();
-
     /** Change listeners. */
     private List<ConfigurationStorageListener> listeners = new CopyOnWriteArrayList<>();
 
     /** Storage version. */
     private AtomicLong version = new AtomicLong(0);
 
+    /** Start key in ragne for searching local configuration keys. */
+    private ByteArray localKeysStartRange = ByteArray.fromString(LOCAL_PREFIX + ".");
+
+    /** End key in range for searching local configuration keys. */
+    private ByteArray localKeysEndRange = ByteArray.fromString(LOCAL_PREFIX + (char)('.' + 1));
+
     /** {@inheritDoc} */
     @Override public synchronized Data readAll() throws StorageException {
-        return new Data(new HashMap<>(map), version.get(), 0);
+        Iterator<Entry> iter =
+            vaultMgr.range(localKeysStartRange, localKeysEndRange);
+
+        HashMap<String, Serializable> data = new HashMap<>();
+
+        while (iter.hasNext()) {
+            Entry val = iter.next();
+
+            data.put(val.key().toString().replaceFirst(LOCAL_PREFIX + ".", ""),
+                (Serializable)ByteUtils.fromBytes(val.value()));
+        }
+
+        return new Data(data, version.get());
     }
 
     /** {@inheritDoc} */
-    @Override public synchronized CompletableFuture<Boolean> write(Map<String, Serializable> newValues, long version) {
-        if (version != this.version.get())
+    @Override public synchronized CompletableFuture<Boolean> write(Map<String, Serializable> newValues, long sentVersion) {
+        if (sentVersion != version.get())
             return CompletableFuture.completedFuture(false);
 
+        CompletableFuture[] futs = new CompletableFuture[newValues.size()];
+
+        int i = 0;
+
         for (Map.Entry<String, Serializable> entry : newValues.entrySet()) {
+            ByteArray key = ByteArray.fromString(LOCAL_PREFIX + "." + entry.getKey());
+
             if (entry.getValue() != null)
-                map.put(entry.getKey(), entry.getValue());
+                futs[i++] = vaultMgr.put(key, ByteUtils.toBytes(entry.getValue()));
```

Comment:
I told you why I don't like this solution.

Reply:
Ok, little comment: "put/remove All" operation should be atomic, it never intertwines with other configuration updates.
Not only that, but "put/remove All" should be fault-tolerant. Current code will produce inconsistent data after fail-restart scenario.

Reply:
@sanpwc Could you please clarify, is it true, that it is possible that some configuration keys might be set bypassing `LocalConfigurationStorage#write` method? We discussed with @ibessonov in private and seems that such things should be restricted.

Reply:
Not sure, let's discuss it tomorrow. However local configuration mutators  should still be atomic within the scope of all newValues, otherwise it'll be possible to see partly updated configuration.

Reply:
> Not sure, let's discuss it tomorrow. 
Well, after some thinking, I agree that we shouldn't update cfg-vault keys directly.